pint-xarray blog post (#251)
xarray-contrib/xarray.dev: src/posts/introducing-pint-xarray/index.md

> # Unit-aware arithmetic in xarray, via pint

Reviewer: > @andersy005, do I need to do something special to be able see a log of the failing build? I get 404s when I try to follow the links.

@keewis, we are missing the necessary metadata information... The following changes should fix the issue 

```suggestion
---
title: "Unit-aware arithmetic in Xarray, via pint"
date: "2022-07-06"
authors:
  - Justus Margin
  - Tom Nicholas
summary: "Xarray now supports unit-aware operations by wrapping pint arrays"
---
```

---
title: "Unit-aware arithmetic in Xarray, via pint"
date: "2022-07-06"
authors:
  - Justus Margin
  - Tom Nicholas
summary: "Xarray now supports unit-aware operations by wrapping pint arrays"
---

_TLDR: Xarray now supports unit-aware operations by wrapping [pint arrays](https://pint.readthedocs.io/en/stable/), so your code can automatically track the physical units that your data represents:_

```ipython
In [2]: distance = xr.DataArray(10).pint.quantify("metres")
   ...: time = xr.DataArray(4).pint.quantify("seconds")
   ...:
   ...: distance / time
Out[2]:
<xarray.DataArray ()>
<Quantity(2.5, 'meter / second')>

> ```

Reviewer 2: Looking at the rendered blog post, I think that it's perfectly easy to parse the input and output without the IPython prompts, and we should just accept this change as good enough.

## Units are integral to science

All quantities in science have units, whether explicitly or implicitly. (And even dimensionless quantities like ratios still technically have units.)

Getting our units right is finicky, and can very easily go unnoticed in our code.

> Even worse, the consequences of getting units wrong can be huge!

Reviewer: i'll look into this and will report back.

> Even worse, the consequences of getting units wrong can be huge!

Reviewer: This has been fixed in #292

The most famous example of a units error has to be NASA's $125 million [Mars Climate Orbiter](https://www.simscale.com/blog/2017/12/nasa-mars-climate-orbiter-metric/), which in 1999 burned up in the Martian atmosphere instead of successfully entering orbit around Mars.
A trajectory course correction had gone wrong, and the error was eventually traced back to a units mismatch: the engineers at Lockheed Martin expressed impulse in [pound-force](<https://en.wikipedia.org/wiki/Pound_(force)>) seconds, whereas the engineers at JPL assumed the impulse value their part of the software received was in SI newton seconds.

<p align = "center">
  <img src = "https://clqtg10snjb14i85u49wifbv-wpengine.netdna-ssl.com/wp-content/uploads/2017/12/Customers.jpg" />
</p>

<p align = "center">
  Newspaper cartoon depicting the incongruence in the units used by NASA and Lockheed Martin scientists that led to the Mars Climate Orbiter disaster.
</p>

> We should take stories like this seriously: If we can automatically track units we can potentially eliminate a whole class of possible errors in our scientific work...

Reviewer: fixed!

## Pint tracks units

There are a few packages for handling units in python (notably [unyt](https://github.com/yt-project/unyt) and [astropy.units](https://docs.astropy.org/en/stable/units/)), but for technical reasons we began units integration in xarray with [pint](https://pint.readthedocs.io/en/stable/).
These various packages work by providing a numerical array type that acts similarly to a numpy array, and is intended to plug in and replace the raw numpy array (a so-called "duck array type").

Pint provides the `Quantity` object, which is a normal numpy array combined with a `pint.Unit`:

```python
q = np.array([6, 7]) * pint.Unit('metres')
print(repr(q))
```

```
Out: <Quantity([6 7], 'meter')>
```

> Pint Quantities act like numpy arrays, except that the units are carried around with the arrays, propagated through operations, and checked during operations involving multiple quantities.

Reviewer 3: ```suggestion
Pint Quantities act like NumPy arrays, except that the units are carried around with the arrays, propagated through operations, and checked during operations involving multiple quantities.
```

## Xarray now wraps Pint

> Thanks to the [tireless work](https://github.com/pydata/xarray/issues/3594) of xarray core developer Justus Magin, you can now enjoy this automatic unit-handling in xarray!

Reviewer 3: ```suggestion
Thanks to the [tireless work](https://github.com/pydata/xarray/issues/3594) of Xarray core developer [Justus Magin](https://github.com/keewis), you can now enjoy this automatic unit-handling in Xarray!
```

Reviewer 2: I also think you don't really need to change this either. It's grammatically a bit weird to speak about yourself in the third person, but this way you get due credit for your work!

> Once you create a unit-aware xarray object (see below for how) you can see the units of the data variables displayed as part of the printable representation.

Reviewer 3: ```suggestion
Once you create a unit-aware xarray object (see below for how) you can see the units of the data variables displayed as part of the printable representation.
```

> You also immediately get the key benefits of pint:

Reviewer 3: ```suggestion
You also immediately get the key benefits of Pint:
```

   ```python
   distance = xr.DataArray(10).pint.quantify("metres")
   time = xr.DataArray(4).pint.quantify("seconds")

   distance / time
   ```

   ```
   Out:
   <xarray.DataArray ()>
   <Quantity(2.5, 'meter / second')>
   ```

2. Dimensionally inconsistent units are caught automatically:

   ```python
   apples = xr.DataArray(10).pint.quantify("kg")
   oranges = xr.DataArray(200).pint.quantify("cm^3")

   apples + oranges
   ```

   ```
   Out:
   DimensionalityError: Cannot convert from 'kilogram' ([mass]) to 'centimeter ** 3' ([length] ** 3)
   ```

3. Unit conversions become simple:

   ```python
   walk = xr.DataArray(500).pint.quantify('miles')

   walk.pint.to('parsecs')
   ```

   ```
   Out:
   <xarray.DataArray ()>
   <Quantity(2.6077643524162074e-11, 'parsec')>
   ```

4. You can specify that custom functions should expect certain units, and convert them if needed:

   ```python
   def jpl_trajectory_code(impulse):
       """This function will only compute the correct result if supplied input in units of Newton-seconds."""

       if impulse.pint.units != "newton * second":
           impulse = impulse.pint.to("Newton * seconds")

       # do some rocket science

>        ...

Author: This feels a bit strange to me: `pint` will happily do the computation, converting units along the way to make sure the results stay correct. At the end of the computation, all we have to do is ask to get the desired units. For this particular example:
```python
def jpl_trajectory_code(..., impulse):
    # do trajectory propagation
    ...

    return propagated_position.pint.to("km"), propagated_velocity.pint.to("km/s")
```

For functions that don't support units, we'd either write a wrapper that does the conversion (no need to check if the units already are what we need) and strips the units, then adds units to the result, or use `expects` once that's ready.

Reviewer 2: Feel free to change the details of this example. The point just needs to be that unit-aware code would have prevented the error that befell the JPL mission. Exactly whether the code raises an error, converts, or whatever doesn't really matter for the pedagogical purposes of this blog post.

Author: I just noticed that with that it becomes essentially the same as 3 (unit conversion). Do you want to replace / extend that section? 4 should then be a section on wrapping library functions.

Reviewer 2: Okay fair point that then they would be the same. Instead consider:

The point of section 3 is showing that you can do unit conversions, the point of section 4 is showing that you can require specific units for finicky library functions. The easiest way to demonstrate that you can require specific units on an input is to raise an error. We don't need to over-think it beyond that.

Therefore I now think that this would be fine for the purposes of this blog post

```python
   def jpl_trajectory_code(impulse):
       """A function which will only compute the correct result if supplied input in units of Newton-seconds."""

       if impulse.pint.units != "newton * second":
           raise pint.DimensionalityError()

       # do some rocket science
       ...
```
It gets the point across.

Yes in practice you would probably do this a bit differently, but that's what our pint-xarray docs should demonstrate.

Author: It feels a bit strange to have a blog post advertising `pint` and `pint-xarray` using unidiomatic code... I don't expect this blog post to become as often read as the one on [`kerchunk`](https://medium.com/pangeo/fake-it-until-you-make-it-reading-goes-netcdf4-data-on-aws-s3-as-zarr-for-rapid-data-access-61e33f8fe685), but that [had to be updated](https://github.com/fsspec/kerchunk/issues/170) because people kept using the code from the blog instead of from the documentation.

What if, instead of presenting it as a new feature, we'd show it as an example using 1 to 3? For unit-aware functions, that's what we'd actually end up doing.

Reviewer 2: > What if, instead of presenting it as a new feature, we'd show it as an example using 1 to 3? 

Okay - what would that look like?

Author: the example would be the one from https://github.com/xarray-contrib/xarray.dev/pull/251#discussion_r940312605, but we'd describe it as something like:
```
With these, you can build code that automatically propagates units and converts them where necessary to stay consistent:

<code>

where the final explicit conversion is used to get the desired output units.
```
I'm sure we can improve the wording, though.

Reviewer 2: @keewis I tried implementing your suggestion in https://github.com/xarray-contrib/xarray.dev/pull/251/commits/02838335e696a26edce21965e0fa82bdd70fb9e4, not quite sure if that's what you meant though.

   lockheed_impulse_value = xr.DataArray(5).pint.quantify("force_pounds * seconds")

   jpl_trajectory_code(lockheed_impulse_value)
   ```

   ```
   Out:
   pint.DimensionalityError
   ```

>    (Note: We are adding [new features](https://github.com/xarray-contrib/pint-xarray/pull/143) to make unit specification of function arguments more slick.)

Author: This should probably be something like
```suggestion
Note: We are adding [new features](https://github.com/xarray-contrib/pint-xarray/pull/143) to make specifying the units of  parameters of existing library functions for more slick.
```
now that we are agreeing on making `expects` a way to add unit support to function libraries that don't handle units (essentially, a "unsafe mode")

Reviewer 2: I'm happy with the first suggestion.

> In the abstract, tracking units like this is useful in the same way that labelling dimensions with xarray is useful: it helps us avoid errors by relieving us of the burden of remembering arbitrary information about our data.

Reviewer 3: ```suggestion
In the abstract, tracking units like this is useful in the same way that labelling dimensions with Xarray is useful: it helps us avoid errors by relieving us of the burden of remembering arbitrary information about our data.
```

## Quantifying with pint-xarray

> The easiest way to create a unit-aware xarray object is to use the helper package we made: [pint-xarray](https://github.com/xarray-contrib/pint-xarray).
> Once you `import pint_xarray` you can access unit-related functionality via `.pint` on any xarray DataArray or Dataset (this works via [xarray's accessor interface](https://xarray.pydata.org/en/stable/internals/extending-xarray.html)).

Reviewer 3: ```suggestion
The easiest way to create a unit-aware Xarray object is to use the helper package we made: [pint-xarray](https://github.com/xarray-contrib/pint-xarray).
Once you `import pint_xarray` you can access unit-related functionality via `.pint` on any `DataArray` or `Dataset` (this works via [Xarray's accessor interface](https://xarray.pydata.org/en/stable/internals/extending-xarray.html)).
```

Above we have seen examples of quantifying explicitly, where we specify the units in the call to `.quantify()`.
We can do this for multiple variables too, and we can also pass `pint.Unit` instances:

```python
ds = xr.Dataset({'a': 2, 'b': 10})

ds.pint.quantify({'a': 'kg',
                  'b': pint.Unit('moles')})
```

```
Out:
<xarray.Dataset>
Dimensions:  ()
Data variables:
    a        int64 [kg] 2
    b        int64 [mol] 10
```

Alternatively, we can quantify from the object's `.attrs`, automatically reading the metadata which xarray objects carry around.
If nothing is passed to `.quantify()`, it will attempt to parse the `.attrs['units']` entry for each data variable.

> This means that for scientific datasets which are stored as files with units in their attributes (which netCDF and Zarr can do for example), using pint with xarray becomes as simple as

Reviewer 3: ```suggestion
This means that for scientific datasets which are stored as files with units in their attributes (which netCDF and Zarr can do for example), using Pint with Xarray becomes as simple as:
```

```python
import pint_xarray

ds = open_dataset(filepath).pint.quantify()
```

## Dequantifying

> To convert our pint arrays back into numpy arrays, we can use `.dequantify`.
> This will strip the units from the arrays and replace them into the `.attrs['units']` of each variable.
> This is useful when we want to save our data back to a file, as it means that the current units will be preserved in the attributes of a netcdf file (or zarr store etc.), as long as we just do `ds.pint.dequantify().to_netcdf()`.

Reviewer 3: ```suggestion
To convert our pint arrays back into NumPy arrays, we can use `.dequantify`.
This will strip the units from the arrays and replace them into the `.attrs['units']` of each variable.
This is useful when we want to save our data back to a file, as it means that the current units will be preserved in the attributes of a netCDF file (or Zarr store etc.), as long as we just do `ds.pint.dequantify().to_netcdf(...)`.
```

## Dask integration

> So xarray can wrap dask arrays, and now it can wrap pint quantities… Can we use both together? Yes!

Reviewer 3: ```suggestion
So Xarray can wrap Dask arrays, and now it can wrap Pint quantities… Can we use both together? Yes!
```

> You can get a unit-aware, dask-backed array either by `.pint.quantify()`-ing a chunked array, or you can `.pint.chunk()` a quantified array.

Reviewer 3: ```suggestion
You can get a unit-aware, Dask-backed array either by `.pint.quantify()`-ing a chunked array, or you can `.pint.chunk()` a quantified array.
```

> (If you have dask installed, then `open_dataset(f).pint.quantify()` will already give you a dask-backed, quantified array.)

Reviewer 3: ```suggestion
(If you have Dask installed, then `open_dataset(f).pint.quantify()` will already give you a Dask-backed, quantified array.)
```

> From there you can `.compute()` the dask-backed objects as normal, and the units will be retained.

Reviewer 3: ```suggestion
From there you can `.compute()` the Dask-backed objects as normal, and the units will be retained.
```

(Under the hood we now have an `xarray.DataArray` wrapping a `pint.Quantity`, which wraps a `dask.array.Array`, which wraps a `numpy.ndarray`.

> This "multi-nested duck array" approach can be generalised to include other array libraries (e.g. `scipy.sparse`), but requires [co-ordination](https://github.com/pydata/duck-array-discussion) between the maintainers of the libraries involved.)

Reviewer 3: ```suggestion
This "multi-nested duck array" approach can be generalised to include other array libraries (e.g. `scipy.sparse`), but requires [coordination](https://github.com/pydata/duck-array-discussion) between the maintainers of the libraries involved.)
```

## Unit-aware indexes

> We would love to be able to promote xarray indexes to pint Quantities, as that would allow you to select data subsets in a unit-aware manner like

Reviewer 3: ```suggestion
We would love to be able to promote Xarray indexes to Pint Quantities, as that would allow you to select data subsets in a unit-aware manner like:
```

```python
da = xr.DataArray(name='a', data=[0, 1, 2], dims='x', coords={'x': [1000, 2000, 3000]})
da = da.pint.quantify({'a': 'Pa', 'x': 'm'})

da.pint.sel(x=2 * 'km')
```

> Unfortunately this will not possible until the ongoing work to extend xarray to support [explicit indexes](https://github.com/pydata/xarray/issues/1603) is complete.

Reviewer 3: ```suggestion
Unfortunately this will not possible until the ongoing work to extend Xarray to support [explicit indexes](https://github.com/pydata/xarray/issues/1603) is complete.
```

distance = distance.pint.quantify({'distance': 'metres',
                                   'time': 'seconds'})
print(distance.coords['time'].attrs)
```

```
Out: {'units': <Unit('second')>}
```

This allows us to provide conveniently wrapped versions of common xarray methods like `.sel`, so that you can still select subsets of data in a unit-aware fashion like this:

```python
distance.pint.sel(time=200 * pint.Unit('milliseconds'))
```

```
Out:
<xarray.DataArray 'distance' ()>
<Quantity(20, 'meter')>
Coordinates:
    time     float64 200.0
```

Observe how the `.pint.sel` operation has first converted 200 milliseconds to 0.2 seconds, before finding the distance value that occurs at a time position of 0.2 seconds.

[This wrapping is currently necessary](https://xarray.pydata.org/en/stable/user-guide/duckarrays.html#missing-features) for any operation which needs to be aware of the units of a dimension coordinate of the dataarray, or any xarray operation which relies on an external library (such as calling `scipy` in `.integrate`).

## CF-compliant units for geosciences with cf-xarray

Different fields tend to have different niche conventions about how certain units are defined.

> By default, pint doesn't understand all the unusual units and conventions we use in geosciences.
> But [pint is customisable](https://pint.readthedocs.io/en/stable/defining.html), and with the help of [cf-xarray](https://github.com/xarray-contrib/cf-xarray) we can teach it about these geoscience-specific units.

Reviewer 3: ```suggestion
By default, Pint doesn't understand all the unusual units and conventions we use in geosciences.
But [Pint is customizable](https://pint.readthedocs.io/en/stable/defining.html), and with the help of [cf-xarray](https://github.com/xarray-contrib/cf-xarray) we can teach it about these geoscience-specific units.
```

> If we `import cf_xarray.units` (before `import pint_xarray`) then we can `quantify` example data from a pangeo climate data store :

Reviewer 3: ```suggestion
If we `import cf_xarray.units` (before `import pint_xarray`) then we can `quantify` example climate data from the [Pangeo Project's CMIP6 catalog](https://pangeo-data.github.io/pangeo-cmip6-cloud/):
```

```python
import xarray as xr
import cf_xarray.units  # must come before import pint_xarray
import pint_xarray

ds = xr.open_dataset('gs://cmip6/CMIP6/CMIP/NCAR/CESM2-FV2/historical/r2i1p1f1/Amon/sfcWind/gn/v20200226/', engine='zarr')
ds = ds.pint.quantify()

squared_wind = ds['sfcWind'] ** 2
squared_wind.pint.units
```

```
Out: <Unit('meter ** 2 / second ** 2')>
```

Here (thanks to `cf_xarray`) pint has successfully interpreted the CF-style units `'m s-1'`, then automatically changed them when we squared the wind speed.

## Plotting

> We can complete our real-world example by plotting the data in its new units

Reviewer 3: ```suggestion
We can complete our real-world example by plotting the data in its new units:
```

```python
import cartopy.crs as ccrs
import matplotlib.pyplot as plt

p = squared_wind.isel(time="2014-01").plot(
    subplot_kws=dict(projection=ccrs.Orthographic(-80, 35), facecolor="gray"),
    transform=ccrs.PlateCarree(),
)
p.axes.set_global()
p.axes.coastlines()
plt.show()
```

![cartopy plot of a quantified dataset](/posts/introducing-pint-xarray/squared_wind.png)

> where `xarray.plot` has detected the pint units automatically.

Reviewer 3: ```suggestion
where `xarray.plot` has detected the Pint units automatically.
```

## Conclusion

Please have a go! You will need xarray (v0.20+) and pint-xarray.

> Please also tell us about any bugs you find, or documentation suggestions you have on the [xarray](https://github.com/pydata/xarray/issues) or [pint-xarray issue trackers](https://github.com/xarray-contrib/pint-xarray/issues).
> If you have usage questions you can raise them there, on the [xarray discussions page](https://github.com/pydata/xarray/discussions), or on the [pangeo discourse forum](https://discourse.pangeo.io/).

Reviewer 3: ```suggestion
Please also tell us about any bugs you find, or documentation suggestions you have on the [Xarray](https://github.com/pydata/xarray/issues) or [pint-xarray issue trackers](https://github.com/xarray-contrib/pint-xarray/issues).
If you have usage questions you can raise them there, on the [Xarray discussions page](https://github.com/pydata/xarray/discussions), or on the [Pangeo Discourse forum](https://discourse.pangeo.io/).
```

> The work here to allow xarray to wrap pint objects is part of a [broader effort to generalise xarray](http://xarray.pydata.org/en/stable/roadmap.html#flexible-arrays) to handle a wide variety of data types (so-called "duck array wrapping").
> Along with the incoming [support for flexible indexes](http://xarray.pydata.org/en/stable/roadmap.html#flexible-indexes), we are excited for all the new features that this will enable for xarray users!

Reviewer 3: ```suggestion
The work here to allow Xarray to wrap Pint objects is part of a [broader effort to generalize Xarray](http://xarray.pydata.org/en/stable/roadmap.html#flexible-arrays) to handle a wide variety of data types (so-called "duck array wrapping").
Along with the incoming [support for flexible indexes](http://xarray.pydata.org/en/stable/roadmap.html#flexible-indexes), we are excited for all the new features that this will enable for Xarray users!
```

```python
import xarray as xr
import cf_xarray.units  # must come before import pint_xarray
import pint_xarray

ds = xr.open_dataset('gs://cmip6/CMIP6/CMIP/NCAR/CESM2-FV2/historical/r2i1p1f1/Amon/sfcWind/gn/v20200226/', engine='zarr')
ds = ds.pint.quantify()

squared_wind = ds['sfcWind'] ** 2
squared_wind.pint.units
```

```
Out: <Unit('meter ** 2 / second ** 2')>
```

Here (thanks to `cf_xarray`) pint has successfully interpreted the CF-style units `'m s-1'`, then automatically changed them when we squared the wind speed.

## Plotting

We can complete our real-world example by plotting the data in its new units:

```python
import cartopy.crs as ccrs
import matplotlib.pyplot as plt

p = squared_wind.isel(time="2014-01").plot(
    subplot_kws=dict(projection=ccrs.Orthographic(-80, 35), facecolor="gray"),
    transform=ccrs.PlateCarree(),
)
p.axes.set_global()
p.axes.coastlines()
plt.show()
```

![cartopy plot of a quantified dataset](/posts/introducing-pint-xarray/squared_wind.png)

where `xarray.plot` has detected the Pint units automatically.

## Conclusion

> Please have a go! You will need xarray (v0.20+) and pint-xarray.

Reviewer 2: Your call. Could either go with whenever the last pint-related bug was fixed or just the most recent release.

If we `import cf_xarray.units` (before `import pint_xarray`) then we can `quantify` example data from a pangeo climate data store :

```python
import xarray as xr
import cf_xarray.units  # must come before import pint_xarray
import pint_xarray

ds = xr.open_dataset('gs://cmip6/CMIP6/CMIP/NCAR/CESM2-FV2/historical/r2i1p1f1/Amon/sfcWind/gn/v20200226/', engine='zarr')
ds = ds.pint.quantify()

squared_wind = ds['sfcWind'] ** 2
squared_wind.pint.units
```

```
Out: <Unit('meter ** 2 / second ** 2')>
```

Here (thanks to `cf_xarray`) pint has successfully interpreted the CF-style units `'m s-1'`, then automatically changed them when we squared the wind speed.

## Plotting

We can complete our real-world example by plotting the data in its new units

```python
import cartopy.crs as ccrs
import matplotlib.pyplot as plt

p = squared_wind.isel(time="2014-01").plot(
    subplot_kws=dict(projection=ccrs.Orthographic(-80, 35), facecolor="gray"),
    transform=ccrs.PlateCarree(),
)
p.axes.set_global()
p.axes.coastlines()
plt.show()
```

> ![cartopy plot of a quantified dataset](squared_wind.png)

Reviewer: one more change :) 

```suggestion
![cartopy plot of a quantified dataset](/posts/introducing-pint-xarray/squared_wind.png)
```

Author: thanks, I was about to try looking up what we did while testing the blog section

   ```python
   distance = xr.DataArray(10).pint.quantify("metres")
   time = xr.DataArray(4).pint.quantify("seconds")

   distance / time
   ```

   ```
   Out:
   <xarray.DataArray ()>
   <Quantity(2.5, 'meter / second')>
   ```

2. Dimensionally inconsistent units are caught automatically:

   ```python
   apples = xr.DataArray(10).pint.quantify("kg")
   oranges = xr.DataArray(200).pint.quantify("cm^3")

   apples + oranges
   ```

   ```
   Out:
   DimensionalityError: Cannot convert from 'kilogram' ([mass]) to 'centimeter ** 3' ([length] ** 3)
   ```

3. Unit conversions become simple:

   ```python
   walk = xr.DataArray(500).pint.quantify('miles')

   walk.pint.to('parsecs')
   ```

   ```
   Out:
   <xarray.DataArray ()>
   <Quantity(2.6077643524162074e-11, 'parsec')>
   ```

4. You can specify that custom functions should expect certain units, and convert them if needed:

   ```python
   def jpl_trajectory_code(impulse):
       """This function will only compute the correct result if supplied input in units of Newton-seconds."""

       if impulse.pint.units != "newton * second":

>            impulse = impulse.pint.to("Newton * seconds")

Reviewer 2: ```suggestion
           raise pint.DimensionalityError()
```

       # do some rocket science
       ...

   lockheed_impulse_value = xr.DataArray(5).pint.quantify("force_pounds * seconds")

   jpl_trajectory_code(lockheed_impulse_value)
   ```

   ```
   Out:
   pint.DimensionalityError
   ```

> Note: We are adding [new features](https://github.com/xarray-contrib/pint-xarray/pull/143) to make specifying the units of parameters of existing library functions for more slick.

Reviewer 2: ```suggestion
Note: We are adding [new features](https://github.com/xarray-contrib/pint-xarray/pull/143) to make specifying the units of parameters of existing library functions more slick.
```
